### PR TITLE
interfaces: add ACRN hypervisor support

### PR DESCRIPTION
**Pull request by igorljubuncic**
This PR is to request support for ACRN hypervisor using a custom interface and the ability to mount hugetables.

### CODE REVIEW

File: interfaces/builtin/acrn.go
Comment (anonymouse64) [inline interfaces/builtin/acrn.go:63]: maybe this should be acrn-support to match other such software like `docker-support` etc?

I added the Needs Samuele Review tag mainly to get review of the name of the interface

File: interfaces/builtin/acrn.go
Comment (anonymouse64) [inline interfaces/builtin/acrn.go:67]: seems like this file needs to be formatted with gofmt

File: interfaces/builtin/acrn.go
Comment (anonymouse64) [inline interfaces/builtin/acrn.go:53]:
```suggestion

type acrnInterface struct {
```

File: interfaces/builtin/acrn.go
Comment (anonymouse64) [inline interfaces/builtin/acrn.go:4]:
```suggestion
 * Copyright (C) 2021 Canonical Ltd
```

**Review by anonymouse64 — COMMENTED** [review]
Seems reasonable given our discussion about the re-implementation, but also needs some unit tests, let us know if you need help with that, otherwise you can take a look at something like adb_support_test.go for an example of what to do to test the interface.

File: interfaces/builtin/acrn.go
Comment (igorljubuncic) [reply interfaces/builtin/acrn.go:67]: No problem, can be linted :)

File: interfaces/builtin/acrn.go
Comment (bboozzoo) [inline interfaces/builtin/acrn.go:33]:
```suggestion
	"github.com/snapcore/snapd/interfaces/apparmor"
	"github.com/snapcore/snapd/snap"
```

File: interfaces/builtin/acrn.go
Comment (bboozzoo) [inline interfaces/builtin/acrn.go:36]: So as far as I understand this allows to operate as ACRN Service VM? https://projectacrn.github.io/latest/developer-guides/hld/hld-overview.html#i-o-emulation Can you confirm this?

File: interfaces/builtin/acrn.go
Comment (bboozzoo) [reply interfaces/builtin/acrn.go:64]: `acrn-support`, or maybe if we have a confirmation that the interface is only useful on a service vm, perhaps `acrn-servicevm-support`?

File: interfaces/builtin/acrn.go
Comment (igorljubuncic) [reply interfaces/builtin/acrn.go:36]: Yes, but I also asked the Intel folks to join the conversation and provide any additional insight on their end.

Comment (gvancuts) [reply interfaces/builtin/acrn.go:36]: This is correct, this is what the Device Model uses to communicate with the hypervisor, you can read more about it here:
* https://projectacrn.github.io/latest/developer-guides/hld/hld-devicemodel.html#hsm
* https://www.kernel.org/doc/html/latest/virt/acrn/index.html

File: interfaces/builtin/acrn.go
Comment (gvancuts) [reply interfaces/builtin/acrn.go:64]: This is correct, this interface is only useful in the Service VM.

File: interfaces/builtin/acrn.go
Comment (pedronis) [inline interfaces/builtin/acrn.go:36]: do we need this bit? cpu-control also provides this afaict?

File: interfaces/builtin/acrn.go
Comment (pedronis) [reply interfaces/builtin/acrn.go:50]: yes a name ending with "-support" is correct for this: will we need other interfaces to support other components? kubernetes-support solves that by having one interface but letting specifying an optional flavor, with that kind of approach "acrn-support" is fine. If basically the service vm is the only thing to support tanyway hen "acrn-support" is maybe also fine.

File: interfaces/builtin/acrn.go
Comment (pedronis) [reply interfaces/builtin/acrn.go:22]: something more high-level like what we use for kurbenetes/docker-support is probably preferable here:

`allows operating as ...`

**Review by pedronis — COMMENTED** [review]
some considerations about naming and other things I noticed. I didn't look at tests.

File: interfaces/builtin/acrn.go
Comment (igorljubuncic) [reply interfaces/builtin/acrn.go:36]: This limits acrn cpu control only to this particular bit. Giving them more is unnecessary security surface.

**Comment by pedronis** [comment]
@igorljubuncic does this supersede #10289?

**Comment by igorljubuncic** [comment]
Yes, it does, as the previous PR stalled due to CLA issues.

**Comment by gvancuts** [comment]
What are the next steps for this PR? I see some checks failing which I guess needs to be addressed, but is there anything else beyond that? Thanks!

**Comment by gvancuts** [comment]
> > What are the next steps for this PR?
> 
> the main thing is a security team review. We can try to prioritize that if needed.

Yes please.

File: interfaces/builtin/acrn.go
Comment (alexmurray) [inline interfaces/builtin/acrn.go:31]: Does this interface need to be super-privileged? It is not clear to me from the documentation whether interacting with the ACRN HSM would allow a snap to cause the hypervisor to make changes to the service VM itself, or does this only allow to control the user VMs? If this does allow to change the service VM from within the service VM then I think it should be super privileged. In that case the snap would need to also declare `deny-connection: true` in `acrnBaseDeclarationSlots`.

**Review by alexmurray — APPROVED** [review]
In general this LGTM but the name of the interface looks like it still needs to be decided as well as whether it should be super privileged or not.

File: interfaces/builtin/acrn.go
Comment (gvancuts) [reply interfaces/builtin/acrn.go:31]: Hi @alexmurray , not sure how to answer your question as I don't know what you mean by "making changes to the Service VM itself". That interface (via the character device) is used to control the lifecycle of User VMs, but there are a number of other operations that the `acrn-dm` will perform in addition to talking to the hypervisor via that character device.

Comment (alexmurray) [reply interfaces/builtin/acrn.go:31]: Yes I understand that the interface is designed to control the User VMs - I was also wondering if perhaps it (intentionally or not) also permitted to control the Service VM as well?

Comment (gvancuts) [reply interfaces/builtin/acrn.go:31]: Hi @alexmurray , I'm asking our engineers/architects to confirm that since I'm not 100% sure of the answer and I wouldn't want to mislead you.

Comment (ywan170) [reply interfaces/builtin/acrn.go:31]: Through the HSM driver, all hypercalls can only control the user VMs. It should not impact the service VM itself. The resource of user VM can be controlled by service VM or hypervisor, but the service VM itself is only can be controlled by hypervisor.

Comment (alexmurray) [reply interfaces/builtin/acrn.go:31]: Ok it sounds like this does not need to be super-privileged then - thanks for the extra info.

Comment (ywan170) [reply interfaces/builtin/acrn.go:31]: The super-privileged means root permission of the hsm device node(cdev), right?

Comment (alexmurray) [reply interfaces/builtin/acrn.go:31]: No, super-privileged means that to use a particular interface, a snap requires a manual declaration be issued by the administrator of the particular store where the snap is uploaded to. Some interfaces essentially allow a snap to take ownership of a device (e.g. the `snapd-control` or `docker-support` interfaces both grant so many permissions to a snap that it can administrer the device the snap is running on). I was trying to ascertain whether this proposed interface could be similar but from the description it is not (since it doesn't provide any abilities to modify / control the VM which the snap is running inside, only the User VMs).

**Comment by gvancuts** [comment]
Hello! What are the next steps to get this finalized and merged?

**Comment by anonymouse64** [comment]
@gvancuts this is waiting on a response to @alexmurray's question in the PR

**Comment by anonymouse64** [comment]
Sorry actually re-reading the discussion this is okay, the interface does not need to be super-privileged so Alex has finished the security approval.

The CI failures were too long ago to still be able to see them, so I'm not sure what those failures were, someone needs to merge master to this branch and see where the tests are, we can help with getting the tests working if needed

**Comment by pedronis** [comment]
I now merged master into this

**Comment by gvancuts** [comment]
> I now merged master into this

Thanks! :-)

**Comment by gvancuts** [comment]
Can someone clarify what is missing to move this along (and hopefully merge it at some point)?

File: tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml
Comment (gvancuts) [inline tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml]: Apologies if this is a silly question but shouldn't that be `acrn-support` since you changed the name of the interface?

File: interfaces/builtin/acrn_support_test.go
Comment (mardy) [inline interfaces/builtin/acrn_support_test.go:4]: nitpick: 2022

File: interfaces/builtin/acrn_support_test.go
Comment (mardy) [inline interfaces/builtin/acrn_support_test.go:101]: nitpick: I would remove the comment and just check
```suggestion
	c.Assert(spec.Snippets()[0], Matches, `^KERNEL=="acrn_hsm", TAG+="snap_consumer_app"`)
```
(untested :-) )

**Review by mardy — APPROVED** [review]
LGTM, thanks!

**Review by pedronis — APPROVED** [review]
pushed some small tweaks, +1

**Pull request merged**
